lib/Runtime/ByteCode/ByteCodeEmitter.cpp

```diff
@@ -6321,8 +6343,122 @@ void EmitDestructuredArray(
+}
+
+// Generating
+// finally {
+//    if (hasReturnFunction)
+//        EmitTryCatchAroundClose
+// }
+
+void EmitFinallyForClose(
+    Js::RegSlot iteratorLocation,
+    Js::RegSlot returnLocation,
+    Js::ByteCodeLabel finallyLabel,
+    ByteCodeGenerator *byteCodeGenerator,
+    FuncInfo *funcInfo)
+{
+    Js::ByteCodeLabel afterFinallyBlockLabel = byteCodeGenerator->Writer()->DefineLabel();
+    byteCodeGenerator->Writer()->Empty(Js::OpCode::Leave);
+
+    byteCodeGenerator->Writer()->RecordCrossFrameEntryExitRecord(false);
+    byteCodeGenerator->Writer()->RecordCrossFrameEntryExitRecord(true);
+
+    byteCodeGenerator->Writer()->Br(afterFinallyBlockLabel);
+    byteCodeGenerator->Writer()->MarkLabel(finallyLabel);
+
+    // Insert finally's body
+    Js::ByteCodeLabel skipCallTryCatchLabel = byteCodeGenerator->Writer()->DefineLabel();
+
+    byteCodeGenerator->Writer()->BrReg1(Js::OpCode::BrFalse_A, skipCallTryCatchLabel, returnLocation);
+    EmitTryCatchAroundClose(iteratorLocation, skipCallTryCatchLabel, byteCodeGenerator, funcInfo);
+
+    byteCodeGenerator->Writer()->MarkLabel(skipCallTryCatchLabel);
+
+    byteCodeGenerator->Writer()->RecordCrossFrameEntryExitRecord(false);
+    byteCodeGenerator->Writer()->Empty(Js::OpCode::LeaveNull);
+    byteCodeGenerator->Writer()->MarkLabel(afterFinallyBlockLabel);
+}
+
+// Emit a wrapper try..finaly block around the destructuring elements
+void EmitDestructuredArray(
+    ParseNode *lhs,
+    Js::RegSlot rhsLocation,
+    ByteCodeGenerator *byteCodeGenerator,
+    FuncInfo *funcInfo)
+{
+    byteCodeGenerator->StartStatement(lhs);
+    Js::RegSlot iteratorLocation = funcInfo->AcquireTmpRegister();
+
+    EmitGetIterator(iteratorLocation, rhsLocation, byteCodeGenerator, funcInfo);
+
+    Assert(lhs->nop == knopArrayPattern);
+    ParseNode *list = lhs->sxArrLit.pnode1;
+
+    if (list == nullptr)
+    { // Handline this case ([] = obj);
+        EmitIteratorClose(iteratorLocation, byteCodeGenerator, funcInfo);
+
+        // No elements to bind or assign.
+        funcInfo->ReleaseTmpRegister(iteratorLocation);
+        byteCodeGenerator->EndStatement(lhs);
+        return;
     }
+
+    // This variable facilitates on when to call the return function (which is Iterator close). When we are emitting bytecode for destructuring element
+    // this variable will be set to true.
+    Js::RegSlot shouldCallReturnFunctionLocation = funcInfo->AcquireTmpRegister();
+    byteCodeGenerator->Writer()->Reg1(Js::OpCode::LdFalse, shouldCallReturnFunctionLocation);
+
+    byteCodeGenerator->SetHasFinally(true);
+    byteCodeGenerator->SetHasTry(true);
+    byteCodeGenerator->TopFuncInfo()->byteCodeFunction->SetDontInline(true);
+
+    // Insert try node here 
+    Js::ByteCodeLabel finallyLabel = byteCodeGenerator->Writer()->DefineLabel();
+    byteCodeGenerator->Writer()->RecordCrossFrameEntryExitRecord(true);
+
+    byteCodeGenerator->Writer()->Br(Js::OpCode::TryFinally, finallyLabel);
```

Comment:
Note to self : Generator case needs to be handled in the destructuring case as well. I have made those changes in my local branch.